File: backend/app/services/csv_parser.py
from __future__ import annotations

import csv
import io
import re
from dataclasses import dataclass
from datetime import date, datetime


@dataclass
class ParsedTransaction:
    date: date
    raw_description: str
    description: str  # Cleaned: stripped, extra whitespace collapsed, uppercased
    amount_cents: int  # Normalized: negative = out, positive = in
    usaa_category: str | None
    status: str  # "Posted" or "Pending"


def _clean_description(raw: str) -> str:
    """Strip whitespace, collapse multiple spaces, uppercase."""
    cleaned = raw.strip()
    cleaned = re.sub(r"\s+", " ", cleaned)
    return cleaned.upper()


def _dollars_to_cents(amount_str: str) -> int:
    """Convert dollar string to integer cents, handling float precision."""
    return int(round(float(amount_str) * 100))


class UsaaCsvParser:
    """Parser for USAA bank CSV exports.

    USAA CSV format:
        Date,Description,Original Description,Category,Amount,Status

    Sign convention differs by account type:
    - Credit card: charges are POSITIVE, payments/credits are NEGATIVE
    - Checking/Savings: spending is NEGATIVE, income is POSITIVE

    The parser normalizes so ALL accounts use: negative = money out, positive = money in.
    """

    def __init__(self, account_type: str) -> None:
        if account_type not in ("checking", "credit", "savings"):
            msg = f"Unsupported account type: {account_type}"
            raise ValueError(msg)
        self.account_type = account_type

    def parse(self, file_content: bytes, filename: str) -> list[ParsedTransaction]:
        """Parse a USAA CSV file into normalized ParsedTransaction objects.

        Skips rows with Status='Pending'.
        For credit cards, flips the sign (USAA has charges as positive).
        For checking/savings, keeps as-is.
        """
        text = file_content.decode("utf-8-sig")  # Handle BOM if present
        reader = csv.DictReader(io.StringIO(text))

        transactions: list[ParsedTransaction] = []

        for row in reader:
            status = row.get("Status", "").strip()
            if status == "Pending":
                continue

            date_str = row["Date"].strip()
            txn_date = datetime.strptime(date_str, "%Y-%m-%d").date()

            raw_description = row.get("Original Description", "").strip()
            if not raw_description:
                raw_description = row.get("Description", "").strip()

            description = _clean_description(raw_description)

            amount_cents = _dollars_to_cents(row["Amount"].strip())

            # Normalize sign convention
            if self.account_type == "credit":
                # USAA credit: positive = charge (money out), negative = payment (money in)
                # We want: negative = money out, positive = money in
                amount_cents = -amount_cents

            usaa_category = row.get("Category", "").strip() or None
            if usaa_category == "Category Pending":
                usaa_category = None

            transactions.append(
                ParsedTransaction(
                    date=txn_date,
                    raw_description=raw_description,
                    description=description,
                    amount_cents=amount_cents,
                    usaa_category=usaa_category,
                    status=status,
                )
            )

        return transactions

File: frontend/src/lib/local-engine/csv-parser.ts
import type { AccountType } from "./types";

export interface ParsedTransaction {
  date: string;
  raw_description: string;
  description: string;
  amount_cents: number;
  usaa_category: string | null;
}

function cleanDescription(raw: string): string {
  return raw.trim().replace(/\s+/g, " ").toUpperCase();
}

function dollarsToCents(amountStr: string): number {
  return Math.round(parseFloat(amountStr) * 100);
}

function parseCsvRows(text: string): Record<string, string>[] {
  const lines = text.split(/\r?\n/);
  if (lines.length < 2) return [];

  const headerLine = lines[0];
  const headers = parseCSVLine(headerLine);

  const rows: Record<string, string>[] = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const values = parseCSVLine(line);
    const row: Record<string, string> = {};
    for (let j = 0; j < headers.length; j++) {
      row[headers[j].trim()] = (values[j] ?? "").trim();
    }
    rows.push(row);
  }
  return rows;
}

function parseCSVLine(line: string): string[] {
  const result: string[] = [];
  let current = "";
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    if (inQuotes) {
      if (char === '"') {
        if (i + 1 < line.length && line[i + 1] === '"') {
          current += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        current += char;
      }
    } else {
      if (char === '"') {
        inQuotes = true;
      } else if (char === ",") {
        result.push(current);
        current = "";
      } else {
        current += char;
      }
    }
  }
  result.push(current);
  return result;
}

export function parseUsaaCsv(
  fileContent: string,
  accountType: AccountType
): ParsedTransaction[] {
  const rows = parseCsvRows(fileContent);
  const transactions: ParsedTransaction[] = [];

  for (const row of rows) {
    const status = row["Status"] ?? "";
    if (status === "Pending") continue;

    const dateStr = (row["Date"] ?? "").trim();
    if (!dateStr) continue;

    let rawDescription = (row["Original Description"] ?? "").trim();
    if (!rawDescription) {
      rawDescription = (row["Description"] ?? "").trim();
    }

    const description = cleanDescription(rawDescription);
    let amountCents = dollarsToCents(row["Amount"] ?? "0");

    if (accountType === "credit") {
      amountCents = -amountCents;
    }

    let usaaCategory: string | null = (row["Category"] ?? "").trim() || null;
    if (usaaCategory === "Category Pending") {
      usaaCategory = null;
    }

    transactions.push({
      date: dateStr,
      raw_description: rawDescription,
      description,
      amount_cents: amountCents,
      usaa_category: usaaCategory,
    });
  }

  return transactions;
}
